File: fenwick-tree/fentree.py
from sys import stdout, stdin

LSOne = lambda b: b&(-b)

class FenwickTree:

    def __init__(self, n):
        self.n = n+1
        self.ft = [0 for i in range(n+1)]

    def rsq_util(self, b):
        s = 0

        while b > 0:
            s += self.ft[b]
            b -= LSOne(b)

        return s

    def rsq(self, a, b):
        return self.rsq_util(b) - (0 if a == 1 else self.rsq_util(a-1))

    def adjust(self, k, v):

        while k < self.n:
            self.ft[k] += v
            k += LSOne(k)

if __name__ == "__main__":
    
    n = int(stdin.readline())

    values = list(map(int, stdin.readline().strip().split()))
    
    fenwick_tree = FenwickTree(n)

    for k, v in enumerate(values):
        fenwick_tree.adjust(k+1, v)

        
    q = int(stdin.readline())

    for i in range(q):
        q_u, l_i, r_x = list(map(str, stdin.readline().strip().split()))
        l_i = int(l_i)
        r_x = int(r_x)
        if q_u == 'q':
            stdout.write(str(fenwick_tree.rsq(l_i, r_x))+'\n')
        else:
            fenwick_tree.adjust(l_i, r_x)

File: fenwick-tree/fentree.cpp
#include <iostream>
#include <vector>

#define endl '\n'
#define lli long long int
#define D(x) cout << #x << " = " << (x) << endl;

using namespace std;

typedef vector<lli> vi;

class FenwickTree{
public:
    vi ft;

    FenwickTree(int n){
        ft.assign(n+1, 0);
    }

    lli rsq_util(int b)
    {
        lli sum = 0;
        for(; b; b-= b&(-b)) sum += ft[b];

        return sum;
    }

    lli rsq(int a, int b)
    {
        if(a == 1) return rsq_util(b);
        
        return rsq_util(b) - rsq_util(a-1);
    }

    void adjust(int k, int v){
        while(k < ft.size()){
            ft[k] += v;
            k += k&(-k);
        }
    }
    
};


int main(){

    int n, q, l_i, r_x;
    lli v;
    char op;
    cin >> n;
    
    FenwickTree fenwick_tree(n);

    for(int i = 0; i < n; i++){
        cin >> v;
        fenwick_tree.adjust(i+1, v);
    }

    cin >> q;

    while(q--){
        cin.ignore();
        cin >> op >> l_i >> r_x;

        if(op == 'q')
            cout << fenwick_tree.rsq(l_i, r_x) << endl;
        else fenwick_tree.adjust(l_i, r_x);
    }
    return 0;
}
